libs/gltfio/src/ReusableAnimator.cpp

```diff
@@ -0,0 +1,489 @@
+                t = (time - prevTime) / deltaTime;
+            }
+        }
+
+        if (sampler->interpolation == Sampler::STEP) {
+            t = 0.0f;
+        }
+
+        mImpl->applyAnimation(channel, t, prevIndex, nextIndex);
+    }
+}
+
+void ReusableAnimator::updateBoneMatrices() {
+    auto renderableManager = mImpl->renderableManager;
+    auto transformManager = mImpl->transformManager;
+
+    auto update = [=](const SkinVector& skins, BoneVector& boneVector) {
+        for (const auto& skin : skins) {
+            size_t njoints = skin.joints.size();
+            boneVector.resize(njoints);
+            for (const auto& entity : skin.targets) {
+                auto renderable = renderableManager->getInstance(entity);
+                if (!renderable) {
+                    continue;
+                }
+                mat4f inverseGlobalTransform;
+                auto xformable = transformManager->getInstance(entity);
+                if (xformable) {
+                    inverseGlobalTransform = inverse(transformManager->getWorldTransform(xformable));
+                }
+                for (size_t boneIndex = 0; boneIndex < njoints; ++boneIndex) {
+                    const auto& joint = skin.joints[boneIndex];
+                    TransformManager::Instance jointInstance = transformManager->getInstance(joint);
+                    mat4f globalJointTransform = transformManager->getWorldTransform(jointInstance);
+                    boneVector[boneIndex] =
+                            inverseGlobalTransform *
+                            globalJointTransform *
+                            skin.inverseBindMatrices[boneIndex];
+                }
+                renderableManager->setBones(renderable, boneVector.data(), boneVector.size());
+            }
+        }
+    };
+
+    update(mImpl->asset->mSkins, mImpl->boneMatrices);
+}
+
+float ReusableAnimator::getAnimationDuration(size_t animationIndex) const {
+    return mImpl->animations[animationIndex].duration;
+}
+
+const char* ReusableAnimator::getAnimationName(size_t animationIndex) const {
+    return mImpl->animations[animationIndex].name.c_str();
+}
+
+void ReusableAnimator::clear() {
+    mImpl->asset = nullptr;
+    mImpl->entityMap.clear();
+    mImpl->renderableManager = nullptr;
+    mImpl->transformManager = nullptr;
+    mImpl->boneMatrices.clear();
+    mImpl->weights.clear();
+    delete mImpl->morpher;
+    mImpl->morpher = nullptr;
+}
+
+void ReusableAnimator::addAnimatedAsset(FilamentAsset* assetToAnimate) {
+    FFilamentAsset* asset = upcast(assetToAnimate);
+    mImpl->asset = asset;
+    for (auto iter : asset->mNodeMap) {
+        std::string name = iter.first->name;
+        auto entity = iter.second;
+        mImpl->entityMap[name] = entity;
+    }
+    mImpl->renderableManager = &asset->mEngine->getRenderableManager();
+    mImpl->transformManager = &asset->mEngine->getTransformManager();
+    mImpl->morpher = new CpuMorpher(asset);
```

Comment:
I'm considering changing this. The morpher does not change for a model and constructing it is not free (we need to parse the vertex positions and the morph targets). We should pass in the morpher instead of creating a new one.

Reply:
changed, cpu morpher can be managed internally or created and supplied externally